pkg/service/controller/ingressip/controller.go

```diff
@@ -0,0 +1,592 @@
+	// Tracks allocations that will be performed after the initial sync
+	pendingAllocations []*kapi.Service
+
+	// Protects the transition between initial sync and regular processing
+	mu sync.RWMutex
+	// Tracks whether the initial sync has been processed
+	syncProcessed bool
+
+	// changeHandler does the work. It's factored out for unit testing
+	changeHandler func(change *serviceChange) error
+}
+
+type serviceChange struct {
+	key        string
+	oldService *kapi.Service
+}
+
+// NewIngressIPController creates a new IngressIPController.
+// TODO this should accept a shared informer
+func NewIngressIPController(client kclient.ServicesNamespacer, ipNet *net.IPNet, resyncInterval time.Duration) *IngressIPController {
+	ic := &IngressIPController{
+		client:     client,
+		queue:      workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		maxRetries: 10,
+	}
+
+	ic.cache, ic.controller = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
+				return ic.client.Services(kapi.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
+				return ic.client.Services(kapi.NamespaceAll).Watch(options)
+			},
+		},
+		&kapi.Service{},
+		resyncInterval,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				service := obj.(*kapi.Service)
+				glog.V(4).Infof("Adding service %s/%s", service.Namespace, service.Name)
+				ic.enqueueChange(obj, nil)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				service := cur.(*kapi.Service)
+				glog.V(4).Infof("Updating service %s/%s", service.Namespace, service.Name)
+				ic.enqueueChange(cur, old)
+			},
+			DeleteFunc: func(obj interface{}) {
+				service := obj.(*kapi.Service)
+				glog.V(4).Infof("Deleting service %s/%s", service.Namespace, service.Name)
+				ic.enqueueChange(nil, obj)
+			},
+		},
+	)
+
+	ic.changeHandler = ic.processChange
+
+	ic.ipAllocator = ipallocator.NewAllocatorCIDRRange(ipNet, func(max int, rangeSpec string) allocator.Interface {
+		return allocator.NewContiguousAllocationMap(max, rangeSpec)
+	})
+
+	ic.allocationMap = make(map[string]string)
+
+	return ic
+}
+
+func (ic *IngressIPController) enqueueChange(new interface{}, old interface{}) {
+	ic.mu.Lock()
+	defer ic.mu.Unlock()
+
+	change := &serviceChange{}
+
+	if new != nil {
+		key, err := controller.KeyFunc(new)
+		if err != nil {
+			glog.Errorf("Couldn't get key for object %+v: %v", new, err)
+			return
+		}
+		change.key = key
+	}
+
+	if old != nil {
+		change.oldService = old.(*kapi.Service)
+	}
+
+	ic.queue.Add(change)
+}
+
+// Run begins watching and syncing.
+func (ic *IngressIPController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	go ic.controller.Run(stopCh)
+
+	glog.V(4).Infof("Waiting for the initial sync to be completed")
+	for !ic.controller.HasSynced() {
+		select {
+		case <-time.After(SyncProcessedPollPeriod):
```

Comment:
I was thinking something like
case <-time.After(x*time.Minute):
          glog.Fatalf("Failed to sync services...")